docker/Dockerfile

```diff
@@ -1,21 +1,9 @@
-# Builder image
-FROM alpine:latest AS builder
+FROM node:16-alpine as builder
 
-ARG VERSION
-ARG REPOSITORY=directus/directus
-
-# Get runtime dependencies from optional dependencies
-# defined in package.json of Directus API package
 WORKDIR /directus
-RUN apk add --no-cache jq \
-  && wget -O directus-api-package.json "https://raw.githubusercontent.com/${REPOSITORY}/${VERSION}/api/package.json" \
-  && jq '{ \
-  name: "directus-project", \
-  version: "1.0.0", \
-  description: "Directus Project", \
-  dependencies: .optionalDependencies \
-  }' \
-  directus-api-package.json > package.json
+COPY /dist .
```

Comment:
As we discussed in that ticket you linked, we want to keep it installing from NPM. That's the official production distribution, and I'd like to keep it that way (feel free to convince me otherwise).

Reply:
I think the tight coupling with the official NPM registry is not an optimal solution as it makes forking **impossible**, at least without using a completely custom build process which increases risk of mistakes exponentially. I can see why you want to discourage forking - but there are a lot of legitimate reasons to fork. We for example need the virtual columns (don't know if you remember the PR) which can't be implemented as an extension.

Installing from NPM won't prevent people from forking - it just makes the process of doing so more error-prone.
The process i implemented here is completely deterministic (you can't even forget to npm run build before lerna publish - which you can ATM and would cause old artifacts to be published under a new version) - so i really don't see where is the difference between copying that .tgz or downloading that same .tgz from NPM registry?

Reply:
It's not about preventing forking at all, it's open source software so you can do with it whatever you want. 

Also, I would like to stress that there's a subtle yet crucial difference between _discouraging_ and _not encouraging_. I have no problems with people forking the project, though it'll cause _a lot_ more maintenance overhead for the main project if we encourage it publicly (and therefore "officially support" it).

I've been bitten in the ass before by the tiniest differences in environment and build chain causing invisible bugs that are impossible to solve. Relying on the same official distribution across deployment targets alleviates some of that pain 🤷🏻

Reply:
I agree that doing two separate builds would be a bad idea for the reasons stated.
But I'm not doing that!
There is 1 build being done and then a parallel publish to NPM and dockerhub.
![image](https://user-images.githubusercontent.com/6235675/125291680-f348ea00-e321-11eb-9a19-14f1eef78d6f.png)
So there IS no different build chain!

And yes, I agree on _not encouraging_, but i think what we're doing right now IS _discouraging_ because a fork will have to jump through way more hoops than necessary to get it to even build. How about we mention in the documentation that forking is not supported at all (if it breaks - we don't care)?


@WoLfulus You wrote the initial build process - may I ask your opinion on this? I'm open to putting a few more hours into it, but I'd be really happy if we can come up with an official build process that:
- Is stable (everything done in CI except _lerna version_)
- Is simple and understandable (as little custom code as possible)
- Is efficient (no double installing dependencies for "logger")
- Doesn't explode outside the official directus repo

Reply:
> I agree that doing two separate builds would be a bad idea for the reasons stated.
> But I'm not doing that!
> There is 1 build being done and then a parallel publish to NPM and dockerhub.

Awesome! In that case 💯

Are there any downsides to not relying on NPM to install Directus inside the container? Upgrading Directus version (though that would be a new container that replaces the old)? Maybe around installing extensions through NPM (though I suppose that would be a custom image at that point)?

Reply:
I am relying on NPM install, but providing explicit tarballs for _directus_ and all dependencies.
I am not aware of any downsides.

This is how the generated package.json looks:
![image](https://user-images.githubusercontent.com/6235675/125294568-aa466500-e324-11eb-8aad-3f1c20ed4d07.png)
It also includes the optionalDependencies as before.

If we want to have extensions installed by NPM I think that should be an additional layer based on the official image:

```
FROM directus/directus
RUN npm install @directus/extension-a @directus/extension-b https://github.com/cynova/directus-extension-c
```

Reply:
Perfect, so you're still relying on npm for regular dependency management, but just using local file pointers for the bundles. Makes sense 👍🏻